From 4e6f020b734e96f8892975328c107130c3ff7073 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Tue, 11 Jun 2019 14:20:40 -0700
Subject: [PATCH] TRITON-1719 sdc-migrate abort fails if there's no destination
 CN Reviewed by: Orlando Vazquez <orlando@joyent.com> Approved by: Orlando
 Vazquez <orlando@joyent.com>

---
 lib/vm-migration/migrate.js         | 58 ++++++++++++++++++++++++++++-
 lib/workflows/vm-migration/begin.js |  1 +
 package.json                        |  2 +-
 3 files changed, 58 insertions(+), 3 deletions(-)

diff --git a/lib/vm-migration/migrate.js b/lib/vm-migration/migrate.js
index c51a747..485421b 100644
--- a/lib/vm-migration/migrate.js
+++ b/lib/vm-migration/migrate.js
@@ -255,6 +255,7 @@ function migrateVm(req, res, next) {
     assert.object(req.filteredNetworks, 'req.filteredNetworks');
     assert.func(next, 'next');
 
+    var actionIsRunning = true;
     var OK_EARLY_ABORT = true;
     var log = req.log;
     var vm = req.vm;
@@ -461,6 +462,59 @@ function migrateVm(req, res, next) {
             });
         },
 
+        function abort_with_no_target_server_uuid(ctx, cb) {
+            assert.notEqual(action, 'estimate');
+
+            if (action !== 'abort') {
+                cb();
+                return;
+            }
+
+            assert.object(ctx.migrationRecord, 'ctx.migrationRecord');
+
+            // When there is no target_server_uuid, the target instance was not
+            // allocated, which means we simply mark the migration record as
+            // aborted and be done.
+            var record = ctx.migrationRecord;
+            if (record.target_server_uuid) {
+                cb();
+                return;
+            }
+
+            // Add a progress entry.
+            var currentTimestamp = new Date().toISOString();
+            var progressEntry = {
+                type: 'progress',
+                phase: 'abort',
+                state: 'successful',
+                started_timestamp: currentTimestamp,
+                finished_timestamp: currentTimestamp,
+                current_progress: 100,
+                total_progress: 100
+            };
+            record.progress_history.push(progressEntry);
+
+            // Update record.
+            record.finished_timestamp = currentTimestamp;
+            record.state = 'aborted';
+
+            req.app.moray.putVmMigration(record,
+                    function _abortPutVmMigration(err) {
+                if (err) {
+                    cb(err);
+                    return;
+                }
+
+                // Mark that the action is not running (it's already completed).
+                actionIsRunning = false;
+
+                res.send(200, {
+                    migration: record
+                });
+                cb(OK_EARLY_ABORT);
+            });
+        },
+
         function start_create_new_record(ctx, cb) {
             assert.notEqual(action, 'estimate');
 
@@ -553,8 +607,8 @@ function migrateVm(req, res, next) {
         }
 
         // Keep a record of the running (or soon to be running) phase.
-        if (action === 'begin' || action === 'sync' || action === 'switch' ||
-                action === 'abort') {
+        if (actionIsRunning && (action === 'begin' || action === 'sync' ||
+                action === 'switch' || action === 'abort')) {
             req.migrationTask.record.state = 'running';
             MIGRATION_RECORD_FOR_VM_UUID[vm.uuid] = req.migrationTask.record;
             // Clear old progress events.
diff --git a/lib/workflows/vm-migration/begin.js b/lib/workflows/vm-migration/begin.js
index de6b039..8e71cb7 100644
--- a/lib/workflows/vm-migration/begin.js
+++ b/lib/workflows/vm-migration/begin.js
@@ -378,6 +378,7 @@ function allocateServer(job, cb) {
         }
 
         job.vmPayload.server_uuid = server_uuid;
+        record.target_server_uuid = server_uuid;
 
         cb(null, 'VM allocated to Server ' + server_uuid);
     });
diff --git a/package.json b/package.json
index 124d9ff..9dc0846 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "vmapi",
   "description": "VMs API",
-  "version": "9.8.9",
+  "version": "9.8.10",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

